src/CompilerGlobals.h

```diff
@@ -0,0 +1,61 @@
+#ifndef HALIDE_COMPILER_GLOBALS_H
+#define HALIDE_COMPILER_GLOBALS_H
+
+/** \file
+ * Contains a struct that defines all of the Compiler's global state.
+ */
+
+#include <atomic>
+#include <chrono>
+#include <string>
+
+#include "Util.h"
+
+namespace Halide {
+namespace Internal {
+
+/** This struct is designed to contain all of the *mutable* global data
+ * used by the Halide compiler. (Global data that is declared const must
+ * not go here.) */
+struct Globals {
+    struct TickStackEntry {
+        std::chrono::time_point<std::chrono::high_resolution_clock> time;
+        std::string file;
+        int line;
+    };
+
+    // A counter to use in random_float() calls.
+    std::atomic<int> random_float_counter;
+
+    // A counter to use in random_uint() and random_int() calls.
+    std::atomic<int> random_uint_counter;
+
+    // A counter to use in tagging random variables.
+    // Note that this will be reset by Internal::reset_random_counters().
+    std::atomic<int> random_variable_counter;
+
+    // Counters used for the unique_name() utilities.
+    std::atomic<int> unique_name_counters[num_unique_name_counters];
+
+    // The stack for tic/toc utilities
+    std::vector<TickStackEntry> tick_stack;
```

Comment:
There's no point putting this in the struct either. It's not mutable compiler state that it makes sense to reset. In fact resetting it could result in toc() trying to pop an empty vector.

Reply:
Fair.

Reply:
Thinking more on this -- you're not wrong, but if a multithreaded caller used this facility, we could end up with similarly mismatched vectors. Since this is mostly for transitory debugging usage, maybe it's no big deal, but it is a minor footgun waiting for code that wants to drive Halide from multiple threads.